### DISCUSSION TITLE
Does oidc-client support jwt-bearer as an authorization grant type for azure workload identity?

### DISCUSSION THREAD

**Discussion by Sopka** [quarkusio/quarkus#38458] [Q&A]
Consider the following setup:

The quarkus application is running in an azure kubernetes cluster and is registered with the azure oidc provider (EntraID). The pod in which the application runs is configured to use azure workload identity.

As a result the kubernetes cluster "injects" a JWT token into the pod, available at a specific filesystem path `/var/run/secrets/azure/tokens/azure-identity-token`.

This JWT token can be used instead of the usual client-secret to authenticate the client/application against the oidc provider to get an access token ([RTF-7523, Section 2.2](https://www.rfc-editor.org/rfc/rfc7523#section-2.2))
The request to obtain a valid access_token from the oidc using the injected jwt as an assertion parameter instead of the client_secret is documented here: [Access token request with a federated credential](https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-client-creds-grant-flow#third-case-access-token-request-with-a-federated-credential)

The injected token is only valid for a short period of time (default 1h) and only available inside the kubernetes cluster. This increases security, because there is no need to configure a static client secret anymore.

One way would be to implement a simple custom client request filter to exchange the token, check for expiration, handle caching...

But the whole process looks so similar to what has already been implemented for the oidc client filter documented here: [Quarkus OIDC Supported Token Grants](https://quarkus.io/guides/security-openid-connect-client-reference#other-grants). But the static properties configuration is not enough, because the token is dynamic and must be read from the file-system again after expiration.

Any ideas how to re-use the existing oidc client filter implementations in this case?

[a comment by an automated account is not shown]

**Reply by sberyozkin** [reply]
@Sopka It is supported at the token propagation filter level with the help of OIDC client, the incoming and verified JWT access token, before being propagated,  is exchanged, and a returned access token is propagated instead - the downstream targeted endpoint will deal with this token verification, see how it can be configured here:

https://github.com/quarkusio/quarkus/blob/main/extensions/oidc-token-propagation-reactive/deployment/src/test/resources/application.properties

The token propagation itself is enabled with the `@AccessToken` annotation.

I believe it should work, can you have a look please ? The only thing that I can see is missing is an optional caching of the tokens which were returned in exchange to the JWT bearer grant tokens, we can do it for sure, I think it makes sense for prod cases

**Reply by sberyozkin** [reply]
@Sopka Just to clarify that access token which is injected in the file system will be expected to be submitted to Quarkus as a Bearer access token - and then, as I've described above, it will be exchanged and propagated to the downstream service.

I did not quite understand though if it is what you need, i.e, if your only requirement is to exchange the token and then use it locally. In this case you should be able to use a custom `SecurityIdentityAugmentor` and have an injected OidcClient (configured as shown in the above link) that you will use directly to exchange the current token by passing it to `OidcClient.getTokens(Map.of(...))` method and then using the builder to update the current `SecurityIdentiy`.

Let me know if you need some further clarifications, I believe your requirements can be met in Quarkus, we can consider some tweaks if needed.

**Reply by Sopka** [reply]
Currently I am using `@RegisterProvider(OidcClientRequestReactiveFilter.class)` for a REST Client Reactive class to call the microsoft graph api (`@RegisterRestClient(baseUri = "https://graph.microsoft.com")`)

The oidc-client configuration in `application.properties` sets the necessary client_id and client_secret to authenticate against Azure EntraId:

```
quarkus.oidc-client.auth-server-url=https://login.microsoftonline.com/${myapp.tenant.id}/v2.0
quarkus.oidc-client.grant.type=client
quarkus.oidc-client.client-id=*******
quarkus.oidc-client.credentials.client-secret.value=******
```

The reactive filter will take care of requesting an access_token and pass it to the downstream API via the authorization Header.

I still want to query the token in nearly the same way, with a grant type of `client_credentials`, but without the `client_secret` parameter. Instead of the client_secret you provide a parameter of `client_assertion` with the value of an azure provided JWT token and another static parameter `client_assertion_type=urn:ietf:params:oauth:client-assertion-type:jwt-bearer`.

The JWT token is not passed initially as an access_token via an AUTHORIZATION header. This token is only known by the application itself at runtime and is used instead of the usual client_secret.

Current I don't see a way to 

a) Use the existing filter implementations like `OidcClientRequestReactiveFilter` and let it post a request with the aforementioned parameters `client_assertion` and `client_assertion_type`, instead of `client_secret`.
b) The client_assertion parameter itself can also be not configured statically in an application.properties file, because it is a dynamic value that the application needs to reload from the execution environment.

In this specific case the value of client_assertion is mounted in a Kubernetes POD at path `/var/run/secrets/azure/tokens/azure-identity-token`

Actually it would be enough to inherit from one of the default quarkus ReactiveFilter implementations and have a defined way to set the client_assertion parameters. The azure specific logic how to acquire the JWT token from the execution environment should be left to the custom filter class.

@sberyozkin Your first suggestion looked promising, but I do not use another grant_type of jwt, it is still the client_credentals type. The second suggestion was interesting, because at least there I have another getTokens method to set different parameters, but I don't need to augment my security context. The use-case is still like the first option to just pass the access token to the downstream service.

**Reply by sberyozkin** [reply]
@Sopka I see, I think I understand, 

>I still want to query the token in nearly the same way, with a grant type of client_credentials, but without the client_secret parameter. Instead of the client_secret you provide a parameter of client_assertion with the value of an azure provided JWT token and another static parameter client_assertion_type=urn:ietf:params:oauth:client-assertion-type:jwt-bearer.

I believe setting the authentication method to `client_secret_jwt` is what you are asking about ?

See https://quarkus.io/guides/security-openid-connect-client-reference#oidc-client-authentication, let me know if it resolves the issue

**Reply by sberyozkin** [reply]
@Sopka Never mind, I see what should be done. I think the custom filter will be needed given that the Azure provided token is dynamic, this custom filter will indeed have OidcClient injected, as well as the JsonWebToken and then use the 2nd approach

**Reply by sberyozkin** [reply]
I'll see what can be simplified

**Reply by sberyozkin** [reply]
@Sopka FYI, #38513, I did not read carefully, but now it is quite clear...

**Reply by sberyozkin** [reply]
@Sopka FYI, #38541

**Reply by sberyozkin** [reply]
@Sopka #38541 has been merged, if you'd like you can try the snapshots

**Reply by martaisty** [reply]
Hi :wave:  @Sopka Have you found a solution for that? I have the same issue. @sberyozkin #38541 This PR doesn't help much in this case, PR solves the issue when Quarkus app creates a JWT token using a secret/private key/keystore... But in this case, Azure provides a ready-to-use JWT token which should be sent as `client_assertion` parameter. The token is available at a specific path inside a container `/var/run/secrets/azure/tokens/azure-identity-token`.
In my case, I want to use it not only for oidc-client but also for protecting a web-app using oidc. The web-app uses authorization code flow to get an access token. The request body for getting an access token looks like this
```
client_id=<my app's client it>
&scope=<scope>
&code=<code received from Microsoft Entra Id>
&redirect_uri=http%3A%2F%2Flocalhost%2Fmyapp%2F
&grant_type=authorization_code
&code_verifier=ThisIsntRandomButItNeedsToBe43CharactersLong
&client_assertion_type=urn%3Aietf%3Aparams%3Aoauth%3Aclient-assertion-type%3Ajwt-bearer
&client_assertion=<has to be taken as is from /var/run/secrets/azure/tokens/azure-identity-token>
```
More about Microsoft Entra Workload ID: https://learn.microsoft.com/en-us/azure/aks/workload-identity-overview?tabs=java
It seems to me like it is not only Azure-related problems. But in general, if I want to use OIDC federation together with Kubernetes service account token volume projection, I can't find a good way to configure it. Do you have any suggestions on how to achieve this @sberyozkin ?

**Reply by sberyozkin** [reply]
@martaisty I think #38541 is what you need, see https://github.com/quarkusio/quarkus/pull/38541/files#diff-49472ed8351660d1802077357cad210eeaf8e1278b0f65fcc0e63534b191746d

**Reply by sberyozkin** [reply]
@martaisty In that test, the filter also uses a specific client tenant, but if you have a single client only then only provide this ready to use token and configure client to expect an out of band token: https://github.com/quarkusio/quarkus/pull/38541/files#diff-37e331b023b0b13a09bf93bfb3ee083914f93b452df02daf525559eae5385ecdR14

**Reply by sberyozkin** [reply]
@martaisty Looking at your example, do you want it be done as part of the `quarkus-oidc` authorization code flow ? If you use `quarkus-oidc-client` to do the authorization code flow yourself, then it should definitely work, just register a custom filter, but I'll need to check if that works in the `quarkus-oidc` scope

**Reply by sberyozkin** [reply]
@martaisty Please watch #40013 in case you need it for `quarkus-oidc`, if that is implemented the way it is suggested then it will work for both `quarkus-oidc-client` and `quarkus-oidc`. But as I said, if you do use `quarkus-oidc-client` then you can do it right now as shown in the #38541 test filter

**Reply by martaisty** [reply]
Thank you for your answers @sberyozkin Indeed, that was what I needed. By providing `client_id` and `client_assertion` I was able to get it up and running for `quarkus-oidc-client` exactly how it was required (https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-client-creds-grant-flow#third-case-access-token-request-with-a-federated-credential). By the way, don't know why `client_id` wasn't picked up from `quarkus.oidc-client.client-id`, though `quarkus.oidc-client.scopes` was :shrug: 
I was just thinking about a more plug-and-play way, something like providing a path to the token, like `quarkus.oidc-client.credentials.jwt.path` and using `@OidcClientFilter`, but making own OIdc Client Request Filter is good enough :+1:

**Reply by martaisty** [reply]
Speaking of `quarkus-oidc`... Yes @sberyozkin , I would like to have it as a part of the authorization code flow. I don't use `quarkus-oidc-client` to do it myself. I have two different use cases:
1. Service to service call (one service calls another one on behalf of itself) This one was solved by creating custom Oidc Client Request Filter :heavy_check_mark: 
2. Web app calls API on behalf of user. During the authorization code flow a web app gets access token for further propagation to the API. This one is still unresolved, will continue discussion on this at #40013

**Reply by sberyozkin** [reply]
@martaisty Thanks for testing with quarkus-oidc-client. 

>By the way, don't know why client_id wasn't picked up from quarkus.oidc-client.client-id, though quarkus.oidc-client.scopes was 

I think it is redundant, and probably a doc issue, since you are authenticating the client using a JWT bearer token. It probably has the client id as one of its claims. This is why it is currently coded not to incl the client id if you use this form of the client authentication. The fact it works without the client id being passed explicitly for you confirms it is not necessary in this case.

>I was just thinking about a more plug-and-play way, something like providing a path to the token, like quarkus.oidc-client.credentials.jwt.path and using @OidcClientFilter, but making own OIdc Client Request Filter is good enough 👍

Right this is actually the plan, you'd do something like
```
@Endpoint.Type=Token
public client MyFilter implements OidcRequestFilter {
     void filter(..., OidcRequestContextProperties props) {
            props.set("client_assertion", "1234");
     }
}
```
and it will work for both OIDC client and OIDC extensions. It will require recalculating the buffer but the grant requests are typically short sequences, so it will be cheap. May be some other optimizations can be done

**Reply by martaisty** [reply]
> The fact it works without the client id being passed explicitly for you confirms it is not necessary in this case.

@sberyozkin  It doesn't work without client id passed explicitly, as I mentioned I provided `client_id` explicitly through additional parameters
